## Differential expression analysis

The most common question that we have when doing RNA-seq is looking
for gene expression differences between two criteria (e.g. experiment
vs control, patients vs unaffected). One way to do this is to run a
differential expression analysis.

There are numerous methods that have been developed, and the one that
we're using today has been incorporated to many bulk RNA-seq
pipelines: the `DESeq2` R package.

### Experimental setup

We performed CRISPR on a cell line, using two independent guide RNA
against a gene of interest (guideRNA1 and guideRNA2). We also have a
control guide RNA (ctrlGuide).

We generated four biological replicates (independent transfections),
and made RNAseq libraries. They were sequenced on Illumina,
aligned to the human genome,  and gene expression was counted. These
are "raw counts", which are preferred by `DESeq2` (and another heavily
used R package in bulk RNA-seq: `edgeR`) as its input.

If you want the full details on how to use `DESeq`, you can read their
user guide [here](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

While the same algorithms might not be used for single-cell analyses,
the concepts (and inputs) should be similar.

### Installing packages

We will need to install the following packages for the main part of
the exercise: `DESeq2`, `here` and `NMF`. If we get to batch
correction, we will also need the `limma` package.

```{r installPackages}
if(! requireNamespace("DESeq2",quietly=TRUE)){
    if (! requireNamespace("BiocManager", quietly = TRUE)){
        install.packages("BiocManager")
    }
    BiocManager::install("DESeq2")
}

if(! requireNamespace("here",quietly=TRUE)){
  install.packages("here")
}

if(! requireNamespace("NMF",quietly=TRUE)){
   install.packages("NMF")
}

if(! requireNamespace("limma",quietly=TRUE)){
    if (! requireNamespace("BiocManager", quietly = TRUE)){
        install.packages("BiocManager")
    }
    BiocManager::install("limma")
}
```

### Setting up our R environment

Hopefully, you have installed the `here` and `DESeq2` R packages. The following
code will check that it is installed, and will load the libraries if found.

```{r DESeq2Setup}
if(! requireNamespace("here",quietly=TRUE)){
  install.packages("here")
}

library(here)

if(! requireNamespace("DESeq2",quietly=TRUE)){
    if (! requireNamespace("BiocManager", quietly = TRUE)){
        install.packages("BiocManager")
    }
    BiocManager::install("DESeq2")
}

library(DESeq2)
```

First we read in the data from the file. In this exercise, it is a
gzipped file, but typically it is a plain text table.

```{r readCounts}
data  <-  read.table(gzfile(here("differential_analysis", "CRISPR_expt.cntTable.gz")),sep="\t",header=T,row.names=1)
head(data)
```

Then we load some metadata that describes the results. This typically
is a small table with relevant information like experiment, replicate,
disease state, gender etc.

```{r readMetadata}
mdata <- read.table(here("differential_analysis", "CRISPR_expt_metadata.txt"),sep="\t",header=T,row.names=1)
head(mdata)
```

One thing that you MUST make sure is that your metadata rows matches
exactly the layout of your data columns. I.e. if the first sample is
"guideRNA1_rep1", then both the first row of `mdata` and first column
of `data` should be that.

Now we set up the object (called a `DESeqDataSet`) with our
experimental data

```{r DESeqDataSet}
dds <- DESeqDataSetFromMatrix(countData = data, colData = mdata, design = ~ experiment)
cat("\n")
dds$experiment <- relevel(dds$experiment, ref = "control")
dds
```

Let's explain what the commands are doing

We are generating a `DESeqDataSet` from a counts matrix,
where we provide the count data  (`data`), the column data (`mdata`),
and provide a "design" of your analysis (i.e. what you are interested
in). In this case, we are interested to see the changes that are
influenced by the experiment.

The second line of code is explicitly telling the DESeqDataSet object
that the reference (`ref`) for the experiment is the value, "control",
and so we want to compare things relative to control.

Once we have set this up, we can now run `DESeq2` and start the analysis:

```{r runDESeq2}
dds <- DESeq(dds)
dds
```

This will run a series of functions, such as estimating size factors,
dispersion, removing outliers (if required), fitting model, and
testing. Each of these steps are linked to a separate function that
you can tweak if you want, but for more analyses, the default setup
works quite well.

Once `DESeq2` is done with the analysis, it is now time for us to get
the results:

```{r getResults}
res <- results(dds, alpha = 0.05)
summary(res)
```

This generates a table with the following information for each
feature: 

- baseMean: mean "normalized" expression across all samples
- log2FoldChange: log2 fold change relative to your `ref` ("control"
  in our case)
- lfcSE: log fold change standard error
- stat: Wald statistics used by the program
- pvalue: raw p-value from the differential analysis
- padj: p-value adjusted for multiple testing using Benjamini-Hochberg
  (a.k.a. FDR)
  
While this is really useful to see whether any of your genes of
interest changed significantly, a more useful output is to get things
that are "significant".

Here, we are extracting significant gene expression changes using a
False Discovery Rate (FDR) cutoff less than 0.05:

```{r getSigResults}
sig <- res[! is.na(res$padj) & res$padj < 0.05, ]
summary(sig)
head(sig)
```

Now we have a list of genes that are "significantly" different between
our guide RNA experiment and our control.

### Visualizing our data

There are several ways we can visualize our data

1) MA plot

This is a graph that plots the mean expression of your features
(genes) on the x-axis, and the fold change (e.g. guide vs control) on
the y-axis. In most cases, you should be seeing most of the genes
lying on `y = 0`, since we typically don't expect a global change in
gene expression. Deviations from this could indicate some potential
technical noise.

The `plotMA` function is from the `DESeq2` package.

```{r plotMA}
plotMA(res, ylim = c(-5,5))
```

In the graph above, things that are differentially expressed
(based on FDR < 0.05) are colored, while others are gray.

2) Volcano plot

This is another way to show the data, where we plot the log fold
change on the x-axis, and the FDR (-log10) on the y-axis. 

```{r volcanoPlot}
filtered <- res[! is.na(res$padj), ]
log10FDR <- -1 * log10(filtered$padj)
voldata <- data.frame(filtered$log2FoldChange, log10FDR)
colnames(voldata) = c("log2FoldChange","neglog10FDR")
volcolor <- rep("gray", nrow(voldata))
volcolor[filtered$padj < 0.05] = "black"
volcolor[filtered$padj < 0.05 & filtered$log2FoldChange > 1] = "red"
volcolor[filtered$padj < 0.05 & filtered$log2FoldChange < -1] = "blue"

plot(voldata$log2FoldChange, voldata$neglog10FDR, pch = ".", col=volcolor, xlab = "Log2 fold change", ylab = "-log10 (FDR)", ylim = c(0,6), xlim = c(-3,3))
abline(v = 1, col = "red", lty = 2)
abline(v = -1, col = "blue", lty = 2)
abline(h = -log10(0.05), col = "black", lty = 2)
legend("bottomright",c("Significant upregulation", "Significant downregulation"), cex = 0.5, col=c("red","blue"), pch = 20)
```

Here, we can now see the relationship between fold change and FDR. I'm
highlighting genes which have > 2 fold change and FDR < 0.05 with
either red (upregulated) or blue (downregulated), while those that are
significant but < 2 fold are in black. 
Three lines are also plotted: blue = 2 fold decrease, red = 2 fold increase,
black = FDR of 0.05.

There are quite a few genes that show a small fold change (< 2 fold),
but still considered significant, while a bunch of genes that have
larger fold change (> 2 fold) are not significant. This is not
unusual, as this is dependent on how "tightly" these genes are
expressed across samples (e.g. genes would not be considered
significant if the differences between other variables can explain the
variation seen in your experiment)

3) Heatmap

This is a nice way to show gene expression changes across multiple
samples. We'll use a plotting function from the `NMF` package.

```{r heatmapSetup}
library(NMF)
```

One highly recommended thing to do is to first perform a
[variance-stabilizing
transformation](https://en.wikipedia.org/wiki/Variance-stabilizing_transformation)
(In essence, you are removing most
of the zeroes in the dataset, and then "log" transforming). This can
be done in DESeq2:

```{r vst}
vsd <- varianceStabilizingTransformation(dds)
vstcounts <- assay(vsd)
head(vstcounts)
```

For this example, we want to look at the
differentially expressed genes across the samples, so we need to
filter our transformed data for just those genes:

```{r filterSigHits}
sigVst <- vstcounts[match(rownames(sig),rownames(vstcounts)), ]
head(sigVst)
```

Now we need to set up some labels and coloring for the heatmap

```{r heatmapSetup}
expt <- mdata$experiment
gRNA <- mdata$guideID
rep <- mdata$replicate

exptCol <- c("red","green")
names(exptCol) <- c("guideRNA","control")
gRNAcol <- c("darkred","pink","green")
names(gRNAcol) <- c("g349","g394","g419")
repCol <- seq(1,nlevels(as.factor(rep)))

columnAnn <- data.frame(expt,gRNA,rep)

direction <- rep("No change",nrow(sigVst))
direction[sig$log2FoldChange > 0] = "Upregulated"
direction[sig$log2FoldChange < 0] = "Downregulated"

dirCol <- c("red","blue")
names(dirCol) <- c("Upregulated","Downregulated")

rowAnn <- data.frame(direction)

annColor <- list(direction = dirCol, expt = exptCol, gRNA = gRNAcol, rep = repCol)

annColor
```

I tend to sort my gene expression by log2 fold change, rather than use
hierarchical clustering in the heatmap. You just have to make sure you
sort any row labels the same way.

```{r sortVst}
sigVst <- sigVst[order(sig$log2FoldChange), ]
direction <- direction[order(sig$log2FoldChange)]
rowAnn <- data.frame(direction)
head(sigVst)
```

Now we can plot the heatmap:

`aheatmap` is from the `NMF` package. We are plotting on-screen,
as well as to a PNG file called `heatmap.png`

```{r heatmap}
aheatmap(as.matrix(sigVst), color = "-RdBu:256", scale = "row", Rowv = NA, Colv = TRUE, legend = TRUE, cexRow = 0.5, cexCol = 0.1, breaks = 0, annCol = columnAnn, annRow = rowAnn, annColors = annColor)
aheatmap(as.matrix(sigVst), color = "-RdBu:256", scale = "row", Rowv = NA, Colv = TRUE, legend = TRUE, cexRow = 0.5, cexCol = 0.1, breaks = 0, annCol = columnAnn, annRow = rowAnn, annColors = annColor,filename = here("differential_analysis","heatmap.png"))
```

A few things to notice:

- The values are Z-scores, which is a measurement of standard
  deviation from the mean. In this case, the heatmap calculated the
  mean value and standard deviation for each gene across all samples,
  and then colored the heatmap based on deviation from the mean
  (Z-score of +1 means 1 standard deviation above the mean)
- I allowed the program to use hierarchical clustering on the samples,
  but didn't allow it on the genes.
- There are colored labels for each sample (and technically for each
  gene), which could be useful when trying to see if certainly
  annotations (e.g. experiment) cluster.

We could also allow the program to use hierarchical clustering on the
genes to see if there are "groups" of genes that have similar expression across the samples.

```{r heatmap2}
aheatmap(as.matrix(sigVst), color = "-RdBu:256", scale = "row", Rowv = TRUE, Colv = TRUE, legend = TRUE, cexRow = 0.5, cexCol = 0.1, breaks = 0, annCol = columnAnn, annRow = rowAnn, annColors = annColor)
```

What we can see here is that there are subsets of genes that show
differences between the replicates (some that are much higher in
certain replicates than others). This is not completely unexpected, as
this is the dataset (from our last exercise) where we saw some batch
effects in our replicates.

## Other things we can do

- [What if we have more than two conditions?](#multi_condition) 
   - Can we compare different conditions to each other? 
- [How can we handle batch effect?](#batch_effect)

## More than two conditions {#multi_condition}

While many differential analyses are comparing just two conditions
(e.g. experiment vs control), there are cases where there are more
than two conditions that are under investigation. How can we handle
this?

Luckily, `DESeq2` does allow you to do pairwise comparisons between
any two conditions in your dataset, as long as that information is in
the metadata. For example, if you have three different conditions in
your "patient" metadata (e.g. Alzheimers, ALS, Parkinsons), you could
do pairwise comparisons between them with the same dataset (without
having to manually split them up).

In our example, we have two guide RNAs targeting our gene of interest
(g349 and g394). We might be interested to see if there are guide
RNA-specific effects in our experiment that could be either
interesting or confounding.

Just to remind ourselves what we have in our dataset: 

```{r reminder}
head(mdata)
```

As you can see under the "guideID" column, there are three different
guide RNA ID (two experimental and one control). We can tell `DESeq2`
that instead of using "experiment" for comparison, we want to use
"guideID": 

```{r DESeqDataSet_guideRNA}
dds_gRNA <- DESeqDataSetFromMatrix(countData = data, colData = mdata, design = ~ guideID)
cat("\n")
dds_gRNA
```

Note that we have skipped the line where we tell `DESeq2` what the
"reference" should be. This is because we're going to explicit tell
the program what to compare at a later stage.

Now we run DESeq2: 

```{r runDESeq2_guideRNA}
dds_gRNA = DESeq(dds_gRNA)
```

Now, in order to get the results, we need to explicitly tell the
program which pairwise comparison should be made. This is done by the
following line of code: 

```{r g349vg419}
res_g349vg419  <- results(dds_gRNA, contrast = c("guideID", "g349", "g419"), alpha = 0.05)
summary(res_g349vg419)
```

In the above code, we added a `contrast` variable where we gave it the
name of the condition ("guideID"), the "experimental" value ("g349"),
and the reference value ("g419"). This then generates a comparison
between g349 and g419, with g419 as the "control".

We can easily do a comparison comparing the two guideRNA effects too:

```{r g349vg394}
res_g349vg394  <- results(dds_gRNA, contrast = c("guideID", "g349", "g394"), alpha = 0.05)
summary(res_g349vg394)
```

Hopefully, there are very few genes that are popping up, but it would
give us an idea if there are genes that we should be wary of if they
are highly specific to one guide RNA.

## Accounting for batch effects {#batch_effect}

Previously, we had suspected that there are batch effects in our
datasets (replicates 1 and 2 are different from replicates 3 and
4). We can visualize this with a PCA plot:

```{r plotPCA_batch}
plotPCA(vsd, intgroup = "experiment", ntop = 500)

plotPCA(vsd, intgroup = "replicate", ntop = 500)
```

The majority of the variance in the data is explained by batch, this
is not good when we want to see what is biologically
relevent. Thankfully, DESeq2 has a way to handle this effect using a
generalized linear model
([GLM](https://en.wikipedia.org/wiki/Generalized_linear_model)]. 

Like before, we need to create a DESeqDataSet, but this time, we're
adding replicate as a new variable in our model.

```{r DESeqDataSet_GLM}
dds_GLM <- DESeqDataSetFromMatrix(countData = data, colData = mdata, design = ~ replicate + experiment)
cat("\n")
dds_GLM$experiment <- relevel(dds_GLM$experiment, ref = "control")
dds_GLM
```

The way the "design formula" is written is starting with the tilde
(`~`) symbol, followed by the parameters in your metadata that you
want to model (separated by a `+` symbol). The last variable in your
model is the one that you actually want to do the true comparison,
while the others are additional variables that needed to be accounted
for.

We can now run DESeq2 and get the results (using a FDR cutoff of 0.05):

```{r runDESeq2_GLM}
dds_GLM <- DESeq(dds_GLM)
res_GLM <- results(dds_GLM, alpha = 0.05)
summary(res_GLM)
```

You can see that now there are more differentially expressed
genes. This is because we are now accounting for the variation caused
by the different replicates, and adjusting the comparison based on
experimental conditions.

Let's look at the MA and volcano plots:

```{r GLM_plotMA}
plotMA(res_GLM)
```

```{r GLM_volcano}
filtered <- res_GLM[! is.na(res_GLM$padj), ]
log10FDR <- -1 * log10(filtered$padj)
voldata <- data.frame(filtered$log2FoldChange, log10FDR)
colnames(voldata) = c("log2FoldChange","neglog10FDR")
volcolor <- rep("gray", nrow(voldata))
volcolor[filtered$padj < 0.05] = "black"
volcolor[filtered$padj < 0.05 & filtered$log2FoldChange > 1] = "red"
volcolor[filtered$padj < 0.05 & filtered$log2FoldChange < -1] = "blue"

plot(voldata$log2FoldChange, voldata$neglog10FDR, pch = ".", col=volcolor, xlab = "Log2 fold change", ylab = "-log10 (FDR)", ylim = c(0,6), xlim = c(-3,3))
abline(v = 1, col = "red", lty = 2)
abline(v = -1, col = "blue", lty = 2)
abline(h = -log10(0.05), col = "black", lty = 2)
legend("bottomright",c("Significant upregulation", "Significant downregulation"), cex = 0.5, col=c("red","blue"), pch = 20)
```

Compared to the previous volcano plot, we can see the following differences:

1. More genes that are significant
2. Fewer genes with > 2 fold change that are not significant

This is the result of `DESeq2` correcting for the variation within each replicate before doing the next comparison (guideRNA vs control).

We can also do actual batch correction on the normalized count data so
that we can use it for downstream analysis or visualization.
*Note*: `DESeq2` has already accounted for batch in its differential testing, we are just fixing our normalized count tables

```{r batchCorrection}
if(! requireNamespace("limma",quietly=TRUE)){
    if (! requireNamespace("BiocManager", quietly = TRUE)){
        install.packages("BiocManager")
    }
    BiocManager::install("limma")
}

library(limma)

vsd_corr  <- vsd
assay(vsd_corr) <- limma::removeBatchEffect(assay(vsd_corr), vsd_corr$replicate)
```

Let's now check the corrected structure of the data

```{r plotPCA_corrected}
plotPCA(vsd_corr, intgroup = "experiment", ntop = 500)

plotPCA(vsd_corr, intgroup = "replicate", ntop = 500)
```

Where is most of the variance coming from now? 
Hopefully what we are actually interested in!
